File: java/L800/LeetCode0807.java
package L800;

public class LeetCode0807 {
    /**
     * 东西，南北天际线相同
     * 每一个可增加的最大高度不超过东西，南北的最小值
     */
    public int maxIncreaseKeepingSkyline(int[][] grid) {
        int[] across = new int[grid.length];
        int[] vertical = new int[grid[0].length];
        for (int i = 0;i < grid.length;i++){
            across[i] = 0;
            for (int j = 0;j < grid[0].length;j++){
                if (grid[i][j] > across[i]){
                    across[i] = grid[i][j];
                }
            }
        }
        for (int j = 0;j < grid[0].length;j++){
            vertical[j] = 0;
            for (int i = 0;i < grid.length;i++){
                if (grid[i][j] > vertical[j]){
                    vertical[j] = grid[i][j];
                }
            }
        }
        int add = 0;
        for (int i = 0;i < grid.length;i++){
            for (int j = 0;j < grid[0].length;j++){
                add += Math.min(across[i],vertical[j]) - grid[i][j];
            }
        }
        return add;
    }
}

File: cpp/L800/LeetCode0807.cc
#include <algorithm>
#include <vector>
using std::vector;

class Solution {
public:
    int maxIncreaseKeepingSkyline(vector<vector<int>>& grid) {
        int M = grid.size(), N = grid[0].size();
        vector<int> row(N, 0), col(M, 0);
        for (int i{0};i < M;++i) {
            for (int j{0};j < N;++j) {
                row[j] = std::max(row[j], grid[i][j]);
                col[i] = std::max(col[i], grid[i][j]);
            }
        }
        int ans{0};
        for (int i{0};i < M;++i) {
            for (int j{0};j < N;++j) {
                int x{std::min(col[i], row[j])};
                if (x > grid[i][j]) {
                    ans += x - grid[i][j];
                }
            }
        }
        return ans;
    }
};